File: Day27/scion/pkg/transfer/types_test.go
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package transfer

import (
	"testing"
	"time"
)

func TestFileInfo(t *testing.T) {
	fi := FileInfo{
		Path:     "src/main.go",
		FullPath: "/tmp/test/src/main.go",
		Size:     1024,
		Hash:     "sha256:abc123",
		Mode:     "0644",
	}

	if fi.Path != "src/main.go" {
		t.Errorf("unexpected Path: %s", fi.Path)
	}
	if fi.FullPath != "/tmp/test/src/main.go" {
		t.Errorf("unexpected FullPath: %s", fi.FullPath)
	}
	if fi.Size != 1024 {
		t.Errorf("unexpected Size: %d", fi.Size)
	}
	if fi.Hash != "sha256:abc123" {
		t.Errorf("unexpected Hash: %s", fi.Hash)
	}
	if fi.Mode != "0644" {
		t.Errorf("unexpected Mode: %s", fi.Mode)
	}
}

func TestManifest(t *testing.T) {
	m := Manifest{
		Version:     "1.0",
		ContentHash: "sha256:def456",
		Files: []FileInfo{
			{Path: "file1.txt", Size: 100, Hash: "sha256:aaa"},
			{Path: "file2.txt", Size: 200, Hash: "sha256:bbb"},
		},
	}

	if m.Version != "1.0" {
		t.Errorf("unexpected Version: %s", m.Version)
	}
	if m.ContentHash != "sha256:def456" {
		t.Errorf("unexpected ContentHash: %s", m.ContentHash)
	}
	if len(m.Files) != 2 {
		t.Errorf("unexpected file count: %d", len(m.Files))
	}
}

func TestUploadURLInfo(t *testing.T) {
	expires := time.Now().Add(15 * time.Minute)
	info := UploadURLInfo{
		Path:    "file.txt",
		URL:     "https://storage.example.com/file.txt?token=xyz",
		Method:  "PUT",
		Headers: map[string]string{"Content-Type": "application/octet-stream"},
		Expires: expires,
	}

	if info.Path != "file.txt" {
		t.Errorf("unexpected Path: %s", info.Path)
	}
	if info.Method != "PUT" {
		t.Errorf("unexpected Method: %s", info.Method)
	}
	if info.Headers["Content-Type"] != "application/octet-stream" {
		t.Errorf("unexpected Content-Type header")
	}
}

func TestDownloadURLInfo(t *testing.T) {
	info := DownloadURLInfo{
		Path: "file.txt",
		URL:  "https://storage.example.com/file.txt?token=xyz",
		Size: 500,
		Hash: "sha256:abc123",
	}

	if info.Path != "file.txt" {
		t.Errorf("unexpected Path: %s", info.Path)
	}
	if info.Size != 500 {
		t.Errorf("unexpected Size: %d", info.Size)
	}
	if info.Hash != "sha256:abc123" {
		t.Errorf("unexpected Hash: %s", info.Hash)
	}
}

File: Day21/A2ui/a2ui/renderers/web_core/src/v0_9/catalog/types.test.ts
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import assert from "node:assert";
import { describe, it } from "node:test";
import { Catalog, ComponentApi, createFunctionImplementation } from "./types.js";
import { A2uiExpressionError } from "../errors.js";
import { z } from "zod";

describe("Catalog Types", () => {
  it("creates a catalog with functions", () => {
    const mockComponent: ComponentApi = {
      name: "MockComp",
      schema: z.object({}),
    };

    const mockFunc = createFunctionImplementation(
      {
        name: "mockFunc",
        returnType: "string",
        schema: z.object({})
      },
      () => "result"
    );

    const catalog = new Catalog("test-cat", [mockComponent], [mockFunc]);

    assert.strictEqual(catalog.id, "test-cat");
    assert.strictEqual(catalog.components.size, 1);
    assert.strictEqual(catalog.components.get("MockComp"), mockComponent);

    assert.ok(catalog.functions);
    assert.strictEqual(catalog.functions.size, 1);
    assert.strictEqual(catalog.functions.get("mockFunc"), mockFunc);
  });

  it("throws A2uiExpressionError when function is not found", () => {
    const catalog = new Catalog("test-cat", []);
    const ctx = {} as any;

    assert.throws(
      () => catalog.invoker("nonExistent", {}, ctx),
      (err: any) => {
        return err instanceof A2uiExpressionError &&
               err.message.includes("Function not found") &&
               err.expression === "nonExistent";
      }
    );
  });

  it("throws A2uiExpressionError when zod validation fails", () => {
    const mockFunc = createFunctionImplementation(
      {
        name: "test",
        returnType: "string",
        schema: z.object({
          requiredField: z.string()
        })
      },
      () => "result"
    );
    const catalog = new Catalog("test-cat", [], [mockFunc]);
    const ctx = {} as any;

    assert.throws(
      () => catalog.invoker("test", {}, ctx),
      (err: any) => {
        return err instanceof A2uiExpressionError &&
               err.message.includes("Validation failed") &&
               err.expression === "test" &&
               Array.isArray(err.details);
      }
    );
  });
});
